Implements buttons (#25)

srambach opened the pull request:
Implements buttons, excluding an icon button (no design for an icon button yet).

Fixes #25

patternfly-build commented:
Deploy preview for *pf-next* ready!

Built with commit b871bb28631433dd8501daa85e08ea76f54e5587

https://deploy-preview-192--pf-next.netlify.com

mcarrano commented:
Don't seem to be able to view the preview. Page just hangs in loading.

andresgalante commented on src/patternfly/components/Button/examples/buttonTypes.hbs:
```diff
@@ -1,4 +1,18 @@
+    Danger
+{{/button}}
+
+<span style="background-color: black; padding: 1em .5em; margin: 1em;">
```
I think we should avoid this. If we need a dark background let's modify the workspace and not the example.

andresgalante replied:
I'd leave it white on white for now.

andresgalante commented:
@kybaker please make sure this looks as it should

kybaker commented:
@srambach The text shifts to the right on hover for the active state in the preview section. Other than that good to go.

jgiardino commented:
Looks good! I also noticed the text shifting. 

I'm also curious about the following, but don't expect changes for this PR. I'm mainly just curious whether we would have other border radii.
`$pf-global--BorderRadius:        30em !default;`

jgiardino commented:
Looks good! I don't see the text shifting issue anymore.

christianvogt commented:
Do we want to be able to customize the disabled opacity as well? Often people simply adjust opacity as oppose to assign a separate color for disabled.
Also, should we change the cursor to not-allowed for the disabled button?

christianvogt commented:
@srambach the bootstrap link is enough for me to agree that the default cursor is sufficient :)

jgiardino commented on src/patternfly/components/Button/examples/buttonStates.hbs:
```diff
@@ -1,3 +1,58 @@
+    {{#> button btnClass="pf-is-primary pf-is-active" btnAttributes="ariaPressed=\"true\""}}
+        Primary Active
+    {{/button}}
+    {{#> button btnClass="pf-is-primary pf-is-disabled" btnAttributes="disabled=\"true\" tabIndex=\"-1\" ariaDisabled=\"true\"" }}
```
One of the notes about using ARIA in HTML is that [ARIA adds nothing to the default semantics of most HTML](https://www.w3.org/TR/using-aria/#aria-does-nothing). Since `<button>` natively supports `disabled`, then `aria-disabled` would be considered redundant. 

I did a [quick test page](https://codepen.io/jenngiardino/pen/VxrXrr), and tested in VO on mac Safari, NVDA on FF, and JAWS on Chrome. 

- VO skips the button with `disabled` when using the Tab key to navigate, but will place focus on the button with `aria-disabled`. Both buttons are listed in the rotor menu.
- NVDA also skips the button with `disabled`. I'm not as familiar with NVDA, but couldn't find a dialog that lists form elements.
- JAWS does not skip the button with `disabled`. And both buttons are listed in the dialog that lists form elements.

With either screen reader and either attribute, the button is announced as "unavailable" or "dimmed".

It could be argued that `tabindex="-1"` is redundant when using `disabled`. When using the keyboard to navigate without using a screen reader, `disabled` has the same effect as `tabindex="-1"`. Even though readers were inconsistent in how they responded to the `disabled` with regard to keyboard navigation, using `tabindex="-1"` had no effect on the screen readers.

In this case, I'd suggest using `disabled` and removing `tabindex` and `aria-disabled` attributes.

srambach replied:
@jgiardino I had included the `tabIndex` and `ariaDisabled` because of the comments already existing in the documentation section. Are the last two sentences before the table no longer accurate? I'll remove them also if so.

jgiardino replied:
Oh yeah, good catch. So this is what it is now:
> Disabled buttons should include the aria-disabled="true" attribute to indicate the state of the element to assistive technologies.

> The .pf-is-disabled class uses pointer-events: none to try to disable the link functionality of <a>s, but that CSS property is not yet standardized. In addition, even in browsers that do support pointer-events: none, keyboard navigation remains unaffected, meaning that sighted keyboard users and users of assistive technologies will still be able to activate these links. So to be safe, add a tabindex="-1" attribute on these links (to prevent them from receiving keyboard focus) and use custom JavaScript to disable their functionality.

What about something like this?:
> Disabled buttons should include the `disabled` attribute to indicate the state of the element to assistive technologies.

Oooooh, I'm just now realizing in the second paragraph that it's talking about `<a>` elements. :-/ Yesterday I was testing buttons only. I can test with links when I'm home later to see how they compare, but I'm guessing that they don't support [`disabled`](https://www.w3schools.com/tags/att_disabled.asp). 

Is this example page meant to show examples of the button styles being used for both `<button>` and `<a>` elements? If so, then the second paragraph isn't inaccurate—if we want to simulate a disabled button while using `<a>` then we would need to use the `aria-disabled="true"` and `tabindex="-1"` attributes as described in the second paragraph. I'm inclined to say that we should rely on native semantics when using buttons, and supplement this behavior with aria when using `<a>`, but am interested in what others think.

jgiardino commented on src/patternfly/components/Button/examples/buttonStates.hbs:
```diff
@@ -1,3 +1,58 @@
+    {{#> button btnClass="pf-is-secondary pf-is-focus"}}
+        Secondary Focus
+    {{/button}}
+    {{#> button btnClass="pf-is-secondary pf-is-active" btnAttributes="ariaPressed=\"true\""}}
```
I don't think this is the right case for using `aria-pressed`. According to the [description for aria-pressed](https://www.w3.org/TR/wai-aria-1.1/#aria-pressed), the presence of this attribute indicates that it's a toggle button, like the Mute button shown in this example: https://www.w3.org/TR/wai-aria-practices-1.1/examples/button/button.html

I'd suggest removing this attribute altogether, until we're ready to provide examples of toggle buttons.

srambach replied:
@jgiardino Again, I'm following what was in the existing documentation for buttons regarding the active state. Since we are manually setting `pf-is-active` here to demonstrate what the active state looks like, I read it that we should include `aria-pressed`. If I should remove it from the example, is the comment in the documentation still valid to include?

jgiardino replied:
I wouldn't say this statement is wrong to include: 

> An active state button will appear pressed when active. There’s no need to add a class to <button>s as they use a pseudo-class. However, you can still force the same active appearance with .pf-is-active (and include the aria-pressed="true" attribute) should you need to replicate the state programmatically.

My thinking is that if the attribute `aria-pressed` should only be used for toggle buttons, then we should only include it in examples where we are explicitly showing toggle buttons. 

It would probably be good to clarify the documentation so that it's more obvious what the purpose of the attribute is, and say something like:

> However, if you are using a button as a toggle, you can still force the same active appearance with .pf-is-active (and include the aria-pressed="true" attribute) should you need to replicate the state programmatically.

srambach commented:
@jgiardino I made the documentation changes you suggested and added the `-is-` classes to the documentation. I also removed the reference to using `pf-c-button` on an `<a>` because that is not yet defined or shown.

andresgalante commented on src/patternfly/components/Button/examples/buttonStates.hbs:
```diff
@@ -1,3 +1,58 @@
-{{#> button btnClass="pf-is-primary pf-is-active" btnAriaPressed="true"}}Active{{/button}}
-{{#> button btnClass="pf-is-primary pf-is-disabled" btnTabIndex="-1" btnAriaDisabled="true"}}Disabled{{/button}}
\ No newline at end of file
+<div class="pf-l-level">
```
Why is this layout here? there should be no layout information in the component example code.

srambach replied:
That's a good question - I wanted to lay the example out nicely because it's not one, it's many buttons. Is it a rule that you can't do that? Is it a rule that you don't include more than one "thing" per example? If so, buttons is going to be super long. Or I could do something in the middle and put each type with all its states in one example.

andresgalante commented on src/site/pages/guidelines.md:
````diff
@@ -302,6 +302,8 @@ States of a component should be included as a nested element. This includes hove
 }
 ```
 
+Classes that duplicate a state should be named with the prefix `pf-is-`. For example, a class `pf-is-active` can be used to apply the `:active` styling. 
````
ok, good. So if the class doesn't duplicate a state then we can use `-has-`, right? as in `pf-has-gutters`

andresgalante commented:
@jgiardino a random large radius is the only way to do a rounded ended btn without using css shapes and have good support. are you ok with it?

jgiardino commented:
> a random large radius is the only way to do a rounded ended btn without using css shapes and have good support. are you ok with it?

I'm ok with the method of using a really big radius to create the rounded end button.

I think what seems weird to me is that I expect global variables to be values that I would use throughout components, and that the names of the variables make it obvious what the purpose of the variable is. But the following variable doesn't fit nicely into my view of global variables. If I were looking at the list of global variables and saw this variable, I'd have no idea that it's purpose is to create a rounded end.

`$pf-global--BorderRadius: 30em !default;`

But I also don't have any great suggestions, and I don't think this is worth blocking this PR.

andresgalante commented:
@jgiardino you are right, maybe that should be a component variable with a hardcoded magic number instead of a global variable. 

@srambach @jgiardino What do you think?

seanforyou23 commented on src/patternfly/components/Button/examples/buttonStates.hbs:
```diff
@@ -1,3 +1,58 @@
+    {{#> button btnClass="pf-is-primary pf-is-active"}}
+        Primary Active
+    {{/button}}
+    {{#> button btnClass="pf-is-primary pf-is-disabled" btnAttributes="disabled=\"true\" " }}
```
HTML's `disabled` attribute needs a little more than just being set to function as expected with respect to accessibility. Because it's a boolean attribute [1], simply having the attribute present, even without a value assigned at all, will have some effect depending on your combination of OS and screen reader. The problem is that support will be unstable across a variety of user agent, browsers, etc.

For best support, consider using `disabled="disabled"`. In this case, VoiceOver properly announced the "dimmed" state when it came across it.

[1] https://www.w3.org/TR/html5/infrastructure.html#boolean-attributes

Looks great otherwise, nice work!

srambach replied:
OK, I've made this change. Thanks, @seanforyou23 !

srambach commented:
@andresgalante I removed the layouts.
@jgiardino After talking with @andresgalante, we thought the magic number could stand as a global since it will probably be needed in at least 2 places. I added a comment explaining it.

jgiardino commented:
@andresgalante @srambach Yeah I had a similar thought after seeing the Badges issue. I'm fine with that.

andresgalante commented on src/patternfly/components/Button/docs/code.md:
```diff
@@ -4,20 +4,21 @@ Always add a modifier class to add color to the button.
-Disabled buttons should include the `aria-disabled="true"` attribute to indicate the state of the element to assistive technologies.
-
-The `.pf-is-disabled` class uses `pointer-events: none` to try to disable the link functionality of `<a>`s, but that CSS property is not yet standardized. In addition, even in browsers that do support `pointer-events: none`, keyboard navigation remains unaffected, meaning that sighted keyboard users and users of assistive technologies will still be able to activate these links. So to be safe, add ...
+Disabled buttons should include the `disabled` attribute to indicate the state of the element to assistive technologies.
```
@srambach This needs adjustments following @seanforyou23 comment
